source/tutorials/file-sets.md

```diff
@@ -0,0 +1,725 @@
+```
+
+Follow the instructions in the error message, and use `maybeMissing` to create a file set from a path that may not exist (in which case the file set will be empty):
+
+<!-- https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.maybeMissing -->
+
+```{code-block} diff
+:caption: build.nix
+ { stdenv, lib }:
+ let
+   fs = lib.fileset;
+-  sourceFiles = fs.difference ./. ./result;
++  sourceFiles = fs.difference ./. (fs.maybeMissing ./result);
+ in
+```
+
+This now works, using the whole directory since `./result` is not present:
+
+```
+$ nix-build
+trace: /home/user/select (all files in directory)
+this derivation will be built:
+  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
+...
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+Another build attempt will produce a different trace, but the same output path:
+
+```
+$ nix-build
+trace: /home/user/fileset
+trace: - build.nix (regular)
+trace: - default.nix (regular)
+trace: - hello.txt (regular)
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+## Union (explicitly exclude files)
+
+There is still a problem:
+Changing _any_ of the included files causes the derivation to be rebuilt, even though it doesn't depend on those files.
+
+Append an empty line to `build.nix`:
+
+```shell-session
+$ echo >> build.nix
+```
+
+Again, Nix will start from scratch:
+
+```shell-session
+$ nix-build
+trace: /home/user/select
+trace: - default.nix (regular)
+trace: - nix (all files in directory)
+trace: - build.nix (regular)
+trace: - string.txt (regular)
+this derivation will be built:
+  /nix/store/zmgpqlpfz2jq0w9rdacsnpx8ni4n77cn-filesets.drv
+...
+/nix/store/6pffjljjy3c7kla60nljk3fad4q4kkzn-filesets
+```
+
+One way to fix this is to use [`unions`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.unions).
+
+Create a file set containing a union of the files to exclude (`fs.unions [ ... ]`), and subtract it (`difference`) from the complete directory (`./.`):
+
+```{code-block} diff
+:caption: build.nix
+ { stdenv, lib }:
+ let
+   fs = lib.fileset;
+-  sourceFiles = fs.difference ./. (fs.maybeMissing ./result);
++  sourceFiles =
++    fs.difference
++      ./.
++      (fs.unions [
++        (fs.maybeMissing ./result)
++        ./default.nix
++        ./build.nix
++        ./nix
++      ]);
+ in
+```
+
+Changing any of the excluded files now doesn't necessarily require a rebuild anymore.
+
+Check it and modify one of the excluded files again:
+
+```
+$ echo >> build.nix
+```
+
+```
+$ nix-build
+trace: /home/user/fileset
+trace: - hello.txt (regular)
+trace: - world.txt (regular)
+/nix/store/ckn40y7hgqphhbhyrq64h9r6rvdh973r-fileset
+```
+
+## Filter
+
+Dealing with a large number of files, independent of their location, can be done programmatically with the [`fileFilter`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.fileFilter) function.
```

Comment:
```suggestion
With the [`fileFilter`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.fileFilter) function, you can find sets of files by criteria.
```

Actually close to set builder notation, for the mathematically inclined.

Reply:
Changing this to "find sets of files satisfying a criteria"